source/SquadLeader/app/src/main/java/com/esri/squadleader/model/Mil2525CMessageLayer.java

```diff
@@ -0,0 +1,207 @@
+        super(DictionaryType.MIL2525C, symbolDictionaryPath);
+        init(xmlMessageFilename, name, mapController);
+    }
+
+    private void init(String xmlMessageFilename, String name, MapController mapController) throws ParserConfigurationException, SAXException, IOException {
+        this.setName(name);
+        Mil2525CMessageParser parser = new Mil2525CMessageParser();
+        parser.parseMessages(new File(xmlMessageFilename));
+        final List<Message> messages = parser.getMessages();
+        final MapControllerListener listener = new MapControllerListener() {
+
+            @Override
+            public void mapReady() {
+                synchronized (messages) {
+                    for (Message message : messages) {
+                        try {
+                            //Any other problem simply throws a RuntimeException, but a missing
+                            //message ID crashes the app. Therefore, we test for that case and
+                            //throw our own RuntimeException.
+                            if (null == message.getID()) {
+                                throw new RuntimeException("Message ID is null");
+                            } else {
+                                getMessageProcessor().processMessage(message);
+                            }
+                        } catch (RuntimeException re) {
+                            Logger.getLogger(getClass().getName()).log(Level.WARNING, "Bad message in layer\n\tMessage: " + message + "\n\tError: " + re.getMessage());
+                        }
+                    }
+                    //Now that all the sublayers are created, turn labels on or off.
+                    toggleLabels(isShowLabels());
+                }
+            }
+
+            public void layersChanged(boolean isOverlay) {}
+        };
+        Thread readyThread = new Thread() {
+            @Override
+            public void run() {
+                listener.mapReady();
+            }
+        };
+        if (mapController.isReady()) {
+            readyThread.start();
+        } else {
+            /**
+             * TODO potential race condition. If the map wasn't ready when we checked mapController.isReady()
+             *      but it is ready now, the listener won't get called and the messages won't be processed
+             *      and added to the map. This behavior doesn't seem to happen, probably because some
+             *      other layer is almost always added to the map before this one, but if this becomes
+             *      a problem, figure out how to resolve it.
+             */
+            mapController.addListener(listener);
+        }
+    }
+
+    private void toggleLabels(boolean showLabels) {
+        /**
+         * TODO this is the right way to toggle the labels, but there is a bug in the Runtime SDK for
+         * Android that causes it not to work (NIM102986). When that bug is fixed, uncomment this code
```

Comment:
Might be a good time to check if this bug has been allocated to Quartz - since I believe this issue has been open for a while